compiler/rustc_target/src/spec/tests/tests_impl.rs

```diff
@@ -146,9 +146,6 @@ impl Target {
         if self.position_independent_executables && !triple.ends_with("-linuxkernel") {
             assert_eq!(self.relocation_model, RelocModel::Pic);
         }
-        if self.relocation_model == RelocModel::Pic {
-            assert!(self.dynamic_linking || self.position_independent_executables);
-        }
```

Comment:
This assert should not be removed in general, you can add an exception for UEFI with an explanation instead.

Reply:
done

Reply:
Should uefi set position_independent_executables and static_position_independent_executables?

Reply:
> Should uefi set position_independent_executables and static_position_independent_executables?

Hard to tell by reading the in-code docs what these exactly control. But I think codegen_ssa is the only user and it simply causes PIC output requests to be silently converted to NoPIC output requests. In that case, I think UEFI should set both since PIC is technically supported, though not required nor used by anyone I know.

Reply:
I think this assertion is misleading. UEFI does not support dynamic linking, nor are PIC-executables really necessary. However, it still *requires* a PIC'ish-reloc-model to allow the custom UEFI loaders to relocate the images during load.

I already mentioned this in https://github.com/rust-lang/rust/pull/100537, it is not clear to me what effect `RelocModel::Static` has. The LLVM docs are basically non-existant. If it causes `.reloc` to be stripped from PE32 binaries, then it is not applicable to UEFI.

Reply:
static_position_independent_executables is for when there is no dynamic linker.

> it is not clear to me what effect RelocModel::Static has.

AFAIK it fixes a specific load location at link time (eg 0x10000, depends on the linker) as opposed to allowing it to be loaded anywhere at runtime as with RelocModel::Pic.

Reply:
> static_position_independent_executables is for when there is no dynamic linker.

Well, yeah, sure, but my question was rather what exactly it controls? Does it make rustc refuse compilation of non-pic? does it enforce static-pic regardless of what the caller chose? Does it mean pic is mandatory, or just supported?

Maybe there is no clear definition like this, and it just controls the one knob we currently have in codegen_ssa. That would be fine with me. In that case, UEFI should probably set this to `true`.

> > it is not clear to me what effect RelocModel::Static has.
> 
> AFAIK it fixes a specific load location at link time (eg 0x10000, depends on the linker) as opposed to allowing it to be loaded anywhere at runtime as with RelocModel::Pic.

I see. This likely means on PE32 `.reloc` is stripped and relocation is not supported? In that case, I think the assertion is wrong and needs to go. A relocatable binary should not imply dynamic-linking or position-independen-code. Since reloc and PIC are different things on UEFI (and also on Windows).

Reply:
> Well, yeah, sure, but my question was rather what exactly it controls? Does it make rustc refuse compilation of non-pic? does it enforce static-pic regardless of what the caller chose? Does it mean pic is mandatory, or just supported?

It controls if pic compilation is supported or not. It doesn't make it mandatory. In the linker code for the msvc linker it doesn't distinguish between pic and non-pic executables.

Reply:
I tested adding this into `uefi_msvc_base`:
```
position_independent_executables: true,
static_position_independent_executables: true,
```
and it doesn't seem to cause any problem, at least in the simple test I did. Do you want me to add that to this PR?

Reply:
Friendly ping -- should I add the above lines to this PR? Or should we stick to the existing changes?

Reply:
Not sure.